### ISSUE TITLE
Add server response timeout to GCP Trace

### ISSUE THREAD

**Issue by sergeymeshkov** [spring-cloud/spring-cloud-gcp#1617]
We user Stackdriver Trace and spring-cloud-gcp-starter-trace library. Under high server load, Zipkin throws an error while waiting for server response.
By default AwaitableUnaryClientCallListener timeout 5000 ms and this value can't be changed.
Is it possible to add this configuration to Spring (spring.cloud.gcp.trace*)?

```
Exception in thread "AsyncReporter{StackdriverSender{PROJECT_ID}}" java.lang.IllegalStateException: timeout waiting for onClose. timeoutMs=5000, resultSet=false 
at zipkin2.reporter.stackdriver.internal.AwaitableUnaryClientCallListener.await(AwaitableUnaryClientCallListener.java:45)
at zipkin2.reporter.stackdriver.internal.UnaryClientCall.doExecute(UnaryClientCall.java:46) 
at zipkin2.Call$Base.execute(Call.java:379) 
at zipkin2.Call$Mapping.doExecute(Call.java:237) 
at zipkin2.Call$Base.execute(Call.java:379) 
at zipkin2.reporter.AsyncReporter$BoundedAsyncReporter.flush(AsyncReporter.java:286) 
at zipkin2.reporter.AsyncReporter$Builder$1.run(AsyncReporter.java:190)
```

Issue related to:
https://github.com/openzipkin/zipkin-gcp/issues/121
https://github.com/openzipkin/zipkin-gcp/pull/102

**Comment by meltsufin** [comment]
@sergeymeshkov It's possible to add the configuration property once zipkin-gcp exposes it to us. So, this request is blocked on https://github.com/openzipkin/zipkin-gcp/issues/121#issuecomment-477405256.
Are you planning to raise a PR for zipkin-gcp?

In the meantime, a couple of things that you can try:
1. Increase the number of executor threads: `spring.cloud.gcp.trace.num-executor-threads`.
2. `zipkin2.reporter.stackdriver.internal.AwaitableUnaryClientCallListener.TIMEOUT_MS` actually happens to be exposed as package-private for testing, which you can take advantage of by modifying from a class you put into the same package in your application. I don't recommend it as a long-term solution, but as an emergency workaround, it might work.

**Comment by sergeymeshkov** [comment]
@meltsufin thank you. I planned to create PR, but during my investigation, I noticed the same thing - TIMEOUT_MS is a static package-private variable, which is easily modified in tests. Since we would like to add this value as configurable, it should be replaced to non-static variable. In this case 

`zipkin2.reporter.stackdriver.internal.AwaitableUnaryClientCallListener.TIMEOUT_MS`

**Comment by sergeymeshkov** [comment]
`zipkin2.reporter.stackdriver.internal.AwaitableUnaryClientCallListener.TIMEOUT_MS`

**Comment by sergeymeshkov** [comment]
@meltsufin my apologies, I accidentally closed an issue. I will check implementation and try to create simple workaround PR. For now I will try to increase number `spring.cloud.gcp.trace.num-executor-threads`

**Comment by sergeymeshkov** [comment]
[PR](https://github.com/openzipkin/zipkin-gcp/pull/122)

**Comment by sergeymeshkov** [comment]
@meltsufin PR has been merged to master

**Comment by meltsufin** [comment]
Thanks! Is it in a released version now?
If so, feel free to create a PR here as well to resolve this issue.
We really appreciate contributors like you!

**Comment by sergeymeshkov** [comment]
@meltsufin I'll try to implement this feature right after zipkin release.

**Comment by meltsufin** [comment]
Feel free to stage a PR against the snapshot version in the mean time, if you like.

**Comment by meltsufin** [comment]
We're still not upgrading to the latest version of zipkin-gcp because it's now using a non-GA Trace API.

**Comment by materemias** [comment]
@meltsufin do you have an update on this?

**Comment by meltsufin** [comment]
We've upgraded to the latest version of `zipkin-gcp`. So this issue is no longer blocked.
Contributions are welcome!

**Comment by vafanasy** [comment]
What is the name of the property for setting this timeout? I can't find documentation for this.

We are facing the same issue in our Cloud Run service after adding this dependency `com.google.cloud:spring-cloud-gcp-starter-trace` to the project.

**Comment by meltsufin** [comment]
@vafanasy The configuration property is not exposed yet. I created an issue for it: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/issues/563.

In the meantime, you might be able to just overwrite the `stackdriverSender` bean in your application.

**Comment by elefeint** [comment]
Fixed in GoogleCloudPlatform/spring-cloud-gcp#698, for release in 2.0.6.
